libyul/backends/wasm/EVMToEwasmTranslator.cpp

```diff
@@ -841,18 +844,28 @@ function callvalue() -> z1, z2, z3, z4 {
 	z1, z2, z3, z4 := mload_internal(0:i32)
 }
 function calldataload(x1, x2, x3, x4) -> z1, z2, z3, z4 {
-	eth.callDataCopy(0:i32, u256_to_i32(x1, x2, x3, x4), 32:i32)
+	mstore_internal(0:i32, 0, 0, 0, 0)
+	calldatacopy(0, 0, 0, 0, x1, x2, x3, x4, 0, 0, 0, i64.extend_i32_u(eth.getCallDataSize()))
```

Comment:
calldatacopy should apply the memory offset, shouldn't it?

Reply:
I'm not sure what you exactly mean. Basically I just copy `N` bytes from address (`x1`, `x2`, `x3`, `x4`) to the first memory slot. `N` is defined by the size of the calldata`eth.getCallDataSize()`.

Reply:
You are calling the "non-internal" calldatacopy, this means it will apply a memory offset. Wouldn't it be better to call `eth.callDataCopy` directly?

Reply:
The thing is that if we call `eth.callDataCopy` directly here, overreads are possible - that will lead to a failure in hera. With the "non-internal" calldatacopy function we have already some protection against this. 
But I noticed that the length set here is wrong, it should always limited to be max. 32 bytes, because only 32 bytes are returned.